### PR TITLE
feat: add oneOf in AsyncOperationObject

### PR DESCRIPTION
**Pull request by ChoSeoHwan**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Don't support `oneOf` structure in message field
and override pub/sub messages.

## What is the new behavior?

I want to use [`oneOf` structure in messages field](https://www.asyncapi.com/docs/reference/specification/v2.5.0#operationObject) like below :

### Code 
```typescript
import {
    BaseWsExceptionFilter,
    OnGatewayConnection,
    OnGatewayDisconnect,
    OnGatewayInit,
    SubscribeMessage,
    WebSocketGateway,
    WebSocketServer,
    WsResponse
} from '@nestjs/websockets';
import { AsyncApiPub, AsyncApiService, AsyncApiSub } from 'nestjs-asyncapi';
import * as os from 'os';
import { delay, from, map, Observable, of } from 'rxjs';
import { concatMap } from 'rxjs/operators';
import { Socket, Namespace } from 'socket.io';

import { JoinDto } from '~dto/join-dto';
import { LeaveDto } from '~dto/leave-dto';
import { EmptyDto } from '~dto/empty-dto';
import { PushDto } from '~dto/push-dto';

@AsyncApiService({
    serviceName: 'app'
})
@WebSocketGateway({
    transports: ['websocket'],
    namespace: 'app',
    pingInterval: 1000
})
export class AppGateway
    implements OnGatewayInit, OnGatewayConnection, OnGatewayDisconnect
{
    @WebSocketServer() server: Namespace;
    private logger: Logger = new Logger('AppGateway');

    @SubscribeMessage('join')
    @AsyncApiPub({
        channel: '/app',
        message: {
            name: 'join',
            payload: {
                type: JoinDto
            }
        }
    })
    @AsyncApiSub({
        channel: '/app',
        message: {
            name: 'join',
            payload: {
                type: JoinDto
            }
        }
    })
    async joinEvent(client: Socket, data: JoinDto) {
        // ...
        return {
            event: 'join',
            data
        };
    }

    @SubscribeMessage('leave')
    @AsyncApiPub({
        channel: '/app',
        message: {
            name: 'leave',
            payload: {
                type: LeaveDto
            }
        }
    })
    @AsyncApiSub({
        channel: '/app',
        message: {
            name: 'leave',
            payload: {
                type: LeaveDto
            }
        }
    })
    async leaveEvent(client: Socket, data: JoinDto) {
        const { rooms } = data;
        await Promise.all(rooms.map((room) => client.leave(room)));

        return {
            event: 'leave',
            data: data
        };
    }

    @AsyncApiSub({
        channel: '/app',
        message: {
            name: 'push',
            payload: {
                type: PushDto
            }
        }
    })
    pushEvent(rooms: string[]) {
        this.server.to(rooms).emit('push', {
            server: os.hostname()
        });
    }

    afterInit(server: Namespace) {
        this.logger.log('Init');
    }

    handleDisconnect(client: Socket) {
        this.logger.log(`Client Disconnected : ${client.id}`);
    }

    @AsyncApiSub({
        channel: '/app',
        message: {
            name: 'connected',
            payload: {
                type: EmptyDto
            }
        }
    })
    handleConnection(client: Socket, ...args: unknown[]) {
        this.logger.log(`Client Connected : ${client.id}`);

        client.emit('connected');
    }
}
```

### Expect yaml

```yaml
// ...
channels:
  /app:
    subscribe:
      message:
        oneOf:
          - name: join
            payload:
              $ref: '#/components/schemas/JoinDto'
          - name: leave
            payload:
              $ref: '#/components/schemas/LeaveDto'
          - name: push
            payload:
              $ref: '#/components/schemas/PushDto'
          - name: connected
            payload:
              $ref: '#/components/schemas/EmptyDto'
    publish:
      message:
        oneOf:
          - name: join
            payload:
              $ref: '#/components/schemas/JoinDto'
          - name: leave
            payload:
              $ref: '#/components/schemas/LeaveDto'
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

### CODE REVIEW

**Comment by flamewow** [comment]
hey @ChoSeoHwan, I've left minor comments for the code.
But then I tried to check it out locally and saw no difference.
Try running a sample locally and see if it works for you

**Comment by ChoSeoHwan** [comment]
Sorry for the late reply.

I am not good at English, so please forgive me for using a translator.

Due to the structure of the current project, \
there is a problem that only one decorator is applied when using multiple `AsyncApiPub` or `AsyncApiSub` in the same method.

So, in the current structure, if there is the same `channel` for different methods, `oneOf` is applied.

In addition, I think there is a different part between me and the project manager, so I will tell you this part.

1. I understand that `Channel` corresponds to `Namespace` in `socket.io`.
   - The content described in the [document](https://www.asyncapi.com/docs/concepts/channel) is as follows, \
      which is considered to be consistent with the concept of `Namespace` of `socket.io`.
       > Users can define channels as Topics, Queues, Routing Keys, Routes or Topics depending on the protocol used.
   - However, it seems that the current project uses `channel` as `event`.
  In this case, the question arises how to define multiple `Namespace` s.

2. For the reason of the '1' item, \
   I think the parameters of `AsyncApiService` decorator become descriptions of `channel`, \
   and `AsyncApiPub/AsyncApiSub` should be decorators for messages.

I don't know as much about this project as the project manager, so the above may be wrong.
So, I hope you can share your thoughts on this.

I'm sorry that I'm just raising a problem without contributing much...
However, I really want to use `asyncapi` in a code-first way in nestjs, and I left this answer because this project is my hope.

**Comment by flamewow** [comment]
> Sorry for the late reply.
> 
> I am not good at English, so please forgive me for using a translator.
> 
> Due to the structure of the current project, there is a problem that only one decorator is applied when using multiple `AsyncApiPub` or `AsyncApiSub` in the same method.
> 
> So, in the current structure, if there is the same `channel` for different methods, `oneOf` is applied.
> 
> In addition, I think there is a different part between me and the project manager, so I will tell you this part.
> 
> 1. I understand that `Channel` corresponds to `Namespace` in `socket.io`.
>    
>    * The content described in the [document](https://www.asyncapi.com/docs/concepts/channel) is as follows, 
>      which is considered to be consistent with the concept of `Namespace` of `socket.io`.
>      > Users can define channels as Topics, Queues, Routing Keys, Routes or Topics depending on the protocol used.
>    * However, it seems that the current project uses `channel` as `event`.
>      In this case, the question arises how to define multiple `Namespace` s.
> 2. For the reason of the '1' item, 
>    I think the parameters of `AsyncApiService` decorator become descriptions of `channel`, 
>    and `AsyncApiPub/AsyncApiSub` should be decorators for messages.
> 
> I don't know as much about this project as the project manager, so the above may be wrong. So, I hope you can share your thoughts on this.
> 
> I'm sorry that I'm just raising a problem without contributing much... However, I really want to use `asyncapi` in a code-first way in nestjs, and I left this answer because this project is my hope.

Hey @ChoSeoHwan, 
guess we interpret https://www.asyncapi.com/docs/concepts/channel in different ways.
I see it as event/emit in a socket.io context.

In case you need to define multiple namespaces in a single asyncapi documents I'd define events/emits of that namespace as follows: `${namespaceName}/${eventOrEmitName}` (e.g. 'chat/receiveMessages', 'chat/sendMessage', 'newsFeed/receiveUpdates'). 

The other option is to have separate documents for each namespace (if you have gateways in different modules). I'm working on a major update of the lib that will include a sample that showcases that feature.
Basically, you can create different docs by using 'include' option of the createDocument like:
```
 const doc1 = AsyncApiModule.createDocument(app, asyncApiOptions, {
    include: [ModuleWithNamespace1],
  });

 const doc2 = AsyncApiModule.createDocument(app, asyncApiOptions, {
    include: [ModuleWithNamespace2],
  });
  
 // and then serve those docs on different routes, like so:
  
await AsyncApiModule.setup('/asyncapi-ns1', app, doc1);
await AsyncApiModule.setup('/asyncapi-ns2', app, doc2);
```

**Comment by flamewow** [comment]
@ChoSeoHwan played a bit with your example, guess that's an option as well 😄 

yaml I've used for testing in https://studio.asyncapi.com/

```
asyncapi: '2.5.0'
info:
  title: Account Service
  version: 1.0.0
  description: This service is in charge of processing user signups
channels:
  user/signedup:
    subscribe:
      message:
        oneOf:
            - name: join
              payload:
                type: object
                properties:
                    displayName:
                        type: string
                        description: Name of the user

            - name: leave
              payload:
                type: object
                properties:
                    email:
                        type: string
                        format: email
                        description: Email of the user
```

The only thing that I don't like is that #0, #1 
Screenshot 2022-12-06 at 18 35 16

Gotta figure out how to get rid of that

**Comment by flamewow** [comment]
@ChoSeoHwan the problem I see how to deal with other than message fields in case of such a merge into **oneOf**.
Need to give it more thoughts

**Comment by ChoSeoHwan** [comment]
Thank you for your detailed answer.

I left an answer because there was no way to show it in the documentation for multiple namespaces.

The desired direction is slightly different,
If I proceed in the way you said, I think I will be able to achieve my goal.

In particular, dividing pages by namespace is highly anticipated, and I hope to be able to apply it quickly.

> the problem I see how to deal with other than message fields in case of such a merge into oneOf.
Need to give it more thoughts

According to the direction currently suggested, it seems that there is no need to process overlapping channels.

The most intuitive way to add `oneOf` of the message field is to change the message field of the `@AsyncApiSub/@AsyncApiPub` decorator to receive an array as well,
If it is an array type, it seems to be a way to bind it with oneOf.

To proceed with the above task, 
it seems that you need to modify the fields of the `AsyncOperationOptions` interface in `asyncapi-operation.decorator.ts`
 and the message conversion part in `operation-object.factory.ts`.

However, it is difficult to contribute due to the current company project, 
so I ask for your understanding as it seems difficult for me to edit it myself.
